### ISSUE TITLE
Sameness of ownership transfer barriers

### ISSUE THREAD

**Issue by philiptaylor** [KhronosGroup/Vulkan-Docs#347]
https://www.khronos.org/registry/vulkan/specs/1.0/xhtml/vkspec.html#resources-sharing says:

> To acquire exclusive ownership, the application must execute the same buffer or image memory barrier on a queue from the destination queue family.

What does "the same" mean? I can guess, but I would be happier if the spec specified what it meant.

It can't mean literally the same barrier, because a barrier is a thing that belongs to a single execution of a command buffer on a single queue - it must mean a different barrier with sufficiently similar parameters.

I think it's reasonable to assume they must be the same resource, i.e. buffer/offset/size or image/subresourceRange must have equal values between the barriers; you can't e.g. use different-but-aliased buffers in both, and you can't e.g. release ownership of 2 array layers then acquire ownership of only 1 layer, because that'd be too fiddly to allow.

It also seems reasonable to assume srcQueueFamilyIndex/dstQueueFamilyIndex must have equal values, so that both queues know where it's going to / coming from.

srcLayout/dstLayout are less clear because transitions are unclear, as mentioned in #346.

I guess it's saying srcAccessMask/dstAccessMask must be the same too, but I don't really see why. I don't think an implementation can make use of those values as part of its ownership transfer. I think any barrier like

```
vkCmdPipelineBarrier( // executed on both Q1 and Q2
  srcStageMask=S1,
  dstStageMask=S2,
  pBufferMemoryBarriers = {
    srcAccessMask=A1,
    dstAccessMask=A2,
    srcQueueFamilyIndex=Q1,
    srcQueueFamilyIndex=Q2,
    ...
  },
);
```

could be rewritten to something equivalent like

```
vkCmdPipelineBarrier( // only on Q1
  srcStageMask=S1,
  dstStageMask=S1,
  pBufferMemoryBarriers = {
    srcAccessMask=A1,
    dstAccessMask=0,
    srcQueueFamilyIndex=IGNORED,
    dstQueueFamilyIndex=IGNORED,
  },
);
vkCmdPipelineBarrier( // on both Q1 and Q2
  srcStageMask=S1,
  dstStageMask=S2,
  pBufferMemoryBarriers = {
    srcAccessMask=0,
    dstAccessMask=0,
    srcQueueFamilyIndex=Q1,
    dstQueueFamilyIndex=Q2,
  },
);
vkCmdPipelineBarrier( // only on Q2
  srcStageMask=S2,
  dstStageMask=S2,
  pBufferMemoryBarriers = {
    srcAccessMask=0,
    dstAccessMask=A2,
    srcQueueFamilyIndex=IGNORED,
    dstQueueFamilyIndex=IGNORED,
  },
);
```

which means the transfer barrier can't expect to have any meaningful accessMasks. And if it can't use the accessMasks, it probably shouldn't matter whether they're the same on both queues.

I'm not sure whether the current spec is saying srcStageMask/dstStageMask must be the same - are they part of the barrier or not? They're not part of the VkBufferMemoryBarrier/VkImageMemoryBarrier, but they're part of the defined behaviour of a barrier, so that seems a bit unclear. (And I don't see why they'd need to be the same, for a similar reason to accessMasks.)

**Comment by pimanttr** [comment]
I think I agree that stage and access masks should not matter, because in any case there must be a semaphore or a fence+submit ordering arc between the first barrier on Q1 and the second barrier on Q2 for them to be well-ordered. Besides, Q1 and Q2 may not have the same stage/accesses, which would make this even more iffy.

For the layouts, I think they must be the same. One of the points of ownership transfer is that different queues may implement/interpret layouts differently, and it's not always possible for all queues to transition to a layout that it doesn't know how to interpret. So the actual layout transition may happen in the first barrier, the second barrier, or a combination of both (through a neutral one), hence the need to have the full layout transition information in each barrier.

**Comment by krOoze** [comment]
It's **irrelevant** what we think.
I read it as "identical VkImageMemoryBarrier". It made sense to me to have the same (as in single one identical acting as one) Memory Barrier declared in both queues, so I refrained from reporting this earlier.

But the stages (Execution Dependency) must obviously differ, because of the semaphor in between.

**Comment by oddhack** [comment]
This should be fixed in the 1.0.28 spec update.

**Comment by TobiasHector** [comment]
Sooo... we did close this but actually we came to a *different* conclusion in subsequent discussion. (Sorry, we're only human - we make mistakes :( ). I won't detail exactly what that is right now because the spec change is yet to land, but I will say that the resolution won't break applications doing things the way this issue was resolved (though they may start seeing warnings in the validation layers).

I'll reopen this for now, but make sure that the next change closes this when it goes in. Sorry again to anyone who thought this was done and dusted!

**Comment by krOoze** [comment]
@TobiasHector Forgive my impatience, but can you hint a bit about the _difference_?

From what you now said in #346 and  KhronosGroup/Vulkan-LoaderAndValidationLayers#865 I thought to myself the resolution to this (identical `Vk*MemoryBarrier` structures) is pretty consistent with that planned change.

`stageFlags` also seemed to me relatively straight-forward after the resolution (under usual Vulkan rules for `stageFlags`).

**Comment by TobiasHector** [comment]
@krOoze The proposed edits make it such that *only* the details of the queue families need to be the same, and any layout transfer if one is specified (as per #346). The rest of the details are basically ignored for the relevant *memory* barriers - there's no availability operation defined by a release memory barrier, and no visibility operation defined by an acquire memory barrier.

Whilst I *could* go into more detail about what that really means, to give a complete answer I'd have to explain roughly verbatim a bunch of stuff that is coming in the rewrite, and it's hard to do that without the terminology we're introducing *in* that rewrite - hopefully that is clear enough for now, and you can wait a couple weeks for the rest :smile:?

**Comment by krOoze** [comment]
@TobiasHector No, no, this explanation is quite sufficient for the purpose. Now I understand why the reopen and roughly what to expect... (If I can handle the wait is another matter entirelly though :yum:)

**Comment by oddhack** [comment]
This should be fixed in the 1.0.35 spec update, as part of the synchronization language rewrite.
